### ISSUE TITLE
Maven release packaging issue

### ISSUE THREAD

**Issue by jdimeo** [terminological/r6-generator#11]
Starting to dig in to this sweet plugin!

When I download from Maven central, the parent seems to have been overwritten with a generated/fake POM during the release process. It's not the real one I see here on GitHub, but one with `jar` packaging. I've attached the one that Maven pulled.

[r6-generator-1.1.0-orig.pom.txt](https://github.com/user-attachments/files/17609128/r6-generator-1.1.0-orig.pom.txt)

**Comment by robchallen** [comment]
Sorry for delay replying.

That is the project parent pom. It will get imported along with the individual modules.

You shouldn't need to reference the project directly it was just organised that way for my convenience. It's the modules you need.

The artifacts you want are:

r6-generator-runtime as a runtime dependency
r6-generator-maven-plugin as a build dependency

LMK if that is not enough to go on. Maybe you could post your pom.xml if not.

**Comment by robchallen** [comment]
And yes the parent pom does get rewritten during release to strip out the references to the release mechanics. Is it causing a problem?

**Comment by robchallen** [comment]
Ok I understand now. 
I'll take a look.

**Comment by robchallen** [comment]
I can't replicate this issue. On a fresh user with no maven history I can use the archteype to generate a skeleton project pom and locally install that. The parent pom seems to get downloaded OK:

```XML
test@machine:~/.m2/repository/io/github/terminological/r6-generator/1.1.0$ cat r6-generator-1.1.0.pom
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
        <modelVersion>4.0.0</modelVersion>

        <!--<groupId>io.github.terminological</groupId>-->
        <artifactId>r6-generator</artifactId>
        <packaging>pom</packaging>
        <version>1.1.0</version>
        <name>R6 generator</name>
        <description>
R can use `rJava` to communicate with java. R also has a class system called
R6. If you want to use a java library with native `rJava` in R there is 
potentially a lot of glue code needed, and R library specific packaging 
configuration required. However if you don't mind writing an R-centric API
...
```

There isn't anything in this part of the mechanics that should modify the pom. Only during the build of an R project, with a sources only project does it flatten the POM, but it doesn't sound like you got that far.

Is it worth deleting your `~/.m2/repository/io/github/terminological` directory and trying again?

Could it be some maven configuration issue, on your side?

Mysterious. Happy to jump on a call to try and debug it (maybe on Monday).

**Comment by jdimeo** [comment]
Yeah, thank you so much for looking at this for me. On my work computer, I did already try to remove the past downloads of `io.github.terminological` before (that was the first thing I thought of) and I still hit a snag and had to edit my local cache's copy of the parent POM.

However, I also tried on my personal computer just now and it also seemed to work. Maybe I did a specific weird order on my work computer, like accidentally referencing the parent POM as a `<dependency>` before correcting it to the runtime or maven-plugin, thus confusing Maven to pull the parent as a dependency, and some other CDN caching from Maven made me download the wrong one again.

I think we can close this! I hit another error on my "hello world" return-a-version method from within R, but I'm still working through that and am using your great examples.

**Comment by robchallen** [comment]
Good to hear!

**Comment by robchallen** [comment]
Is there anything I can look at directly? Is the repo private?

**Comment by jdimeo** [comment]
Yes, the repo is private, but here is my POM (it references some of my parent POMs but it shouldn't affect downloading this): 
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<parent>
		<groupId>com.compassion</groupId>
		<artifactId>cof-tools</artifactId>
		<version>5.1.2-SNAPSHOT</version>
		<relativePath>../../pom.xml</relativePath>
	</parent>

	<artifactId>cof-plugin-r</artifactId>
	<name>COF R Plugin</name>
	<description>Plugin that allows COF integration with R.</description>
	
	<properties>
		<maven.deploy.skip>true</maven.deploy.skip>
		<wagon.skip>false</wagon.skip>
		
		<r6.version>1.1.0</r6.version>
		
		<application.file>${project.artifactId}.jar</application.file>
	</properties>
	
	<dependencies>
		<dependency>
			<groupId>${project.groupId}</groupId>
			<artifactId>cof-core</artifactId>
			<version>${cof.core.version}</version>
		</dependency>
		<dependency>
		    <groupId>io.github.terminological</groupId>
		    <artifactId>r6-generator-runtime</artifactId>
		    <version>${r6.version}</version>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>io.github.terminological</groupId>
				<artifactId>r6-generator-maven-plugin</artifactId>
				<version>${r6.version}</version>
				<configuration>
					<outputDirectory>${project.basedir}/..</outputDirectory>
					<packageData>
						<title>COF R Plugin</title>
						<version>0.0.1.9000</version>
						.... all the metadata ...
					</packageData>
				</configuration>
				<executions>
					<execution>
						<id>clean-r-library</id>
						<goals>
							<goal>clean-r-library</goal>
						</goals>
					</execution>
					<execution>
						<id>flatten-pom</id>
						<goals>
							<goal>flatten-pom</goal>
						</goals>
					</execution>
					<execution>
						<id>generate-r-library</id>
						<goals>
							<goal>generate-r-library</goal>
						</goals>
						<phase>package</phase>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
</project>
```

**Comment by robchallen** [comment]
Does it work without the flatten-pom execution in there?

**Comment by robchallen** [comment]
What are your options for these:

						<preCompileBinary>???</preCompileBinary> 
						<packageAllDependencies>???</packageAllDependencies>

**Comment by robchallen** [comment]
Looking at this now I can see that the version of the flatten plugin the code is using is an old one with a different set of options but also a vulnerability. It is possible that this is not getting picked up reliably and hence the options are not being applied correctly. 

I was trying to cater for too many possible use cases to try and make CRAN happy, and its a bit difficult to work out what I was trying to achieve with the pom flattening  :-)

**Comment by jdimeo** [comment]
I didn't specify preCompileBinary but I used true for packageAllDependencies.
Clearing cache still resulted in the same error.

**Comment by robchallen** [comment]
OK. It looks to be something to do with the shade plugin interfering with the flatten plugin.

I think I have replicated the pom packaging issue but it is not throwing an error for me, maybe the original parent pom is cached and it uses that in preference to the shaded version...., but I can see a pom, referencing a parent without the packaging type specified.

I'll take a further look tomorrow, but in the meantime could you try and disable the shade plugin with a 

```
<configuration>
  <packageData>
  ....
     <useShadePlugin>false</useShadePlugin>
  ...

```
and see if the CI works. I'm embarrassed to say I never set up CI on this project, but even if I had I would have struggled to test all the possible paths through the code plus all the possible caching states.

**Comment by jdimeo** [comment]
No rush on this- I'll keep chugging away with local development since I can cache the real parent plugin locally.

**Comment by robchallen** [comment]
Well if its not even fetching the right thing from Maven Central I'm stumped. All I can imagine is that somewhere you are hitting a local maven repository that has done something to the `io.github.terminological:r6-generator` pom - maybe something in your `com.compassion:cof-tools` parent pom. Is it doing a `wagon:merge-maven-repos`, that is somehow substituting the broken pom in?

The pom on Maven Central [here](https://central.sonatype.com/artifact/io.github.terminological/r6-generator) is definitely OK.

**Comment by jdimeo** [comment]
Well the plot thickens. A super minimal repo with just a POM passed on GitHub actions: https://github.com/jdimeo/r6-generator-test
The plugin didn't do anything since there were no files to process, but it successfully downloaded and ran the plugin: https://github.com/jdimeo/r6-generator-test/actions/runs/12550297896/job/34992832740#step:4:16

So this must be something with my internal/private GitLab CI/CD set up...

**Comment by jdimeo** [comment]
This run still failed, but it's some template issue, probably because I didn't specify some bit of metadata that was required. But clearly the plugin correctly ran and generated some stuff in a clean environment: https://github.com/jdimeo/r6-generator-test/actions/runs/12550946137/job/34994517113

**Comment by robchallen** [comment]
> This run still failed, but it's some template issue, probably because I didn't specify some bit of metadata that was required. But clearly the plugin correctly ran and generated some stuff in a clean environment: https://github.com/jdimeo/r6-generator-test/actions/runs/12550946137/job/34994517113

It needed a maintainer email (because CRAN requires this)
I'll take a look at making the optional / mandatory metadata clearer (or some sane defaults or better error messages).

Looking at your CI setup on github I should be able set this up on the `r6-generator-test` project so it is doing the whole workflow.

From the perspective of this issue I think there must be something in the `com.compassion:cof-tools` pom which is either using a maven mirror, or intermediate cache that has an issue with the `io.github.terminolgical:r6-generator` pom, and not sure I can help without direct access to the project.

**Comment by jdimeo** [comment]
Yes, I'm going to close this issue again since I've proven now it's clearly not on your end. Thanks!

**Comment by jdimeo** [comment]
AH!! the Jitpack one is messed up: https://[jitpack.io/io/github/terminological/r6-generator/1.1.0/r6-generator-1.1.0.pom](https://jitpack.io/io/github/terminological/r6-generator/1.1.0/r6-generator-1.1.0.pom)
and my Maven config was pulling that one over the one in Central. We use Jitpack to build some deps that aren't on Central yet and it was getting in the way. :-) Sorry for the rabbit trail here!
